pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go

```diff
@@ -0,0 +1,396 @@
+	Workloads := krt.JoinCollection([]krt.Collection[model.WorkloadInfo]{PodWorkloads, WorkloadEntryWorkloads, ServiceEntryWorkloads}, krt.WithName("Workloads"))
+	return Workloads
+}
+
+func fetchPeerAuthentications(
+	ctx krt.HandlerContext,
+	PeerAuths krt.Collection[*securityclient.PeerAuthentication],
+	meshCfg *MeshConfig,
+	ns string,
+	matchLabels map[string]string,
+) []*securityclient.PeerAuthentication {
+	return krt.Fetch(ctx, PeerAuths, krt.FilterGeneric(func(a any) bool {
+		pol := a.(*securityclient.PeerAuthentication)
+		if pol.Namespace == meshCfg.GetRootNamespace() && pol.Spec.Selector == nil {
+			return true
+		}
+		if pol.Namespace != ns {
+			return false
+		}
+		sel := pol.Spec.Selector
+		if sel == nil {
+			return true // No selector matches everything
+		}
+		return labels.Instance(sel.MatchLabels).SubsetOf(matchLabels)
+	}))
+}
+
+func fetchWaypoints(ctx krt.HandlerContext, Waypoints krt.Collection[Waypoint], ns, sa string) []Waypoint {
+	return krt.Fetch(ctx, Waypoints,
+		krt.FilterNamespace(ns), krt.FilterGeneric(func(a any) bool {
+			w := a.(Waypoint)
+			return w.ForServiceAccount == "" || w.ForServiceAccount == sa
+		}))
+}
+
+func (a *index) constructServicesFromWorkloadEntry(p *networkingv1alpha3.WorkloadEntry, services []model.ServiceInfo) map[string]*workloadapi.PortList {
+	res := map[string]*workloadapi.PortList{}
+	for _, svc := range services {
+		n := namespacedHostname(svc.Namespace, svc.Hostname)
+		pl := &workloadapi.PortList{}
+		res[n] = pl
+		for _, port := range svc.Ports {
+			targetPort := port.TargetPort
+			if named, f := svc.PortNames[int32(port.ServicePort)]; f {
+				// get port name or target port
+				nv, nf := p.Ports[named.PortName]
+				tv, tf := p.Ports[named.TargetPortName]
+				// TODO: is this logic/order correct?
+				if tf {
+					targetPort = tv
+				} else if nf {
+					targetPort = nv
+				} else if named.TargetPortName != "" {
+					// We needed an explicit port, but didn't find one - skip this port
+					continue
+				}
+			}
+			pl.Ports = append(pl.Ports, &workloadapi.Port{
+				ServicePort: port.ServicePort,
+				TargetPort:  targetPort,
+			})
+		}
+	}
+	return res
+}
+
+func workloadNameAndType(pod *v1.Pod) (string, workloadapi.WorkloadType) {
+	objMeta, typeMeta := kubeutil.GetDeployMetaFromPod(pod)
+	switch typeMeta.Kind {
+	case "Deployment":
+		return objMeta.Name, workloadapi.WorkloadType_DEPLOYMENT
+	case "Job":
+		return objMeta.Name, workloadapi.WorkloadType_JOB
+	case "CronJob":
+		return objMeta.Name, workloadapi.WorkloadType_CRONJOB
+	default:
+		return pod.Name, workloadapi.WorkloadType_POD
+	}
+}
+
+func (a *index) constructServices(p *v1.Pod, services []model.ServiceInfo) map[string]*workloadapi.PortList {
+	res := map[string]*workloadapi.PortList{}
+	for _, svc := range services {
+		n := namespacedHostname(svc.Namespace, svc.Hostname)
+		pl := &workloadapi.PortList{}
+		res[n] = pl
+		for _, port := range svc.Ports {
+			targetPort := port.TargetPort
```

Comment:
how is the service TargetPort set, why do we find the targetPort again below?

Reply:
I improved the comments and added a new test. See https://github.com/istio/istio/pull/49162/commits/99ab625274525289167afd36ef5e55425161d01a. Thanks!

I think the logic here was correct just not well described; the workloadEntry part was incorrect, though.

Reply:
Its complex because Service and ServiceEntry behave differently